x/subspaces/ante/fee.go

```diff
@@ -0,0 +1,117 @@
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/cosmos-sdk/x/auth/ante"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+
+	"github.com/desmos-labs/desmos/v4/x/subspaces/types"
+)
+
+var _ sdk.AnteDecorator = &DeductFeeDecorator{}
+
+// DeductFeeDecorator represents the decorator used to deduct fee
+type DeductFeeDecorator struct {
+	authDeductAnte ante.DeductFeeDecorator
+	ak             AccountKeeper
+	bk             BankKeeper
+	sk             SubspacesKeeper
+}
+
+// NewDeductFeeDecorator returns a new DeductFeeDecorator instance
+func NewDeductFeeDecorator(ak AccountKeeper, bk BankKeeper, fk FeegrantKeeper, sk SubspacesKeeper) DeductFeeDecorator {
+	return DeductFeeDecorator{
+		authDeductAnte: ante.NewDeductFeeDecorator(ak, bk, fk),
+		ak:             ak,
+		bk:             bk,
+		sk:             sk,
+	}
+}
+
+// AnteHandle implements AnteDecorator
+func (dfd DeductFeeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (newCtx sdk.Context, err error) {
+	if id, ok := isValidSubspaceTx(tx); ok {
+		feeTx, ok := tx.(sdk.FeeTx)
+		if !ok {
+			return ctx, sdkerrors.Wrap(sdkerrors.ErrTxDecode, "Tx must be a FeeTx")
+		}
+
+		newCtx, success, err := dfd.tryHandleSubspaceTx(ctx, feeTx, id)
+		if err != nil {
+			return newCtx, err
+		}
+		// move to next ante if tryHandleSubspaceTx is success, or using auth.DeductFeeDecorator instead
+		if success {
+			return next(newCtx, tx, simulate)
+		}
+	}
+	return dfd.authDeductAnte.AnteHandle(ctx, tx, simulate, next)
+}
+
+// isValidSubspaceTx returns the valid subspace id, returns false if it is invalid
+func isValidSubspaceTx(tx sdk.Tx) (uint64, bool) {
+	subspaceID := uint64(0)
+	for _, msg := range tx.GetMsgs() {
+		if subspaceMsg, ok := msg.(types.SubspaceMsg); ok {
+			if subspaceID == 0 {
+				subspaceID = subspaceMsg.GetSubspaceID()
+			}
+
+			if subspaceMsg.GetSubspaceID() == subspaceID {
+				continue
+			}
+		}
+
+		return 0, false
+	}
+	return subspaceID, true
+}
+
+// tryHandleSubspaceTx handles the fee deduction for subspace transaction, returns false if the process is failed
+func (dfd DeductFeeDecorator) tryHandleSubspaceTx(ctx sdk.Context, tx sdk.FeeTx, subspaceID uint64) (newCtx sdk.Context, success bool, err error) {
+	fee := tx.GetFee()
+	feePayer := tx.FeePayer()
+	feeGranter := tx.FeeGranter()
+	deductFeesFrom := feePayer
+
+	// if feegranter is not set set or fee granter equals to payer, then use auth.DeductFeeDecorator to deal with fees
+	if feeGranter == nil || feeGranter.Equals(feePayer) {
+		return ctx, false, nil
+	}
+
+	if addr := dfd.ak.GetModuleAddress(authtypes.FeeCollectorName); addr == nil {
+		return ctx, false, fmt.Errorf("fee collector module account (%s) has not been set", authtypes.FeeCollectorName)
+	}
+
+	used := dfd.sk.UseGrantedFees(ctx, subspaceID, feeGranter, feePayer, fee, tx.GetMsgs())
+	if !used {
+		return ctx, false, nil
+	}
+
+	deductFeesFrom = feeGranter
```

Comment:
I just realized another thing, and I would like to start a discussion about this. 

Currently, this way of managing the grants allows one user to grant another user (or user group) the permission to use their funds to pay for fees. This is very similar to the current `x/feegrant` implementation, we're just adding the support for user groups and the spefication of working only for a single subspace. 

However, I imagined the thing to work different for subspaces. In particular, what I think would be more valuable for subspace admins is allowing users to pay for fees using the subspace `treasury` funds. The way I envision this working is: 
1. subspace admins grant a fee allowance to a user or a group 
2. when the user (or group) performs a transaction, if they have a fee allowance, the fee payer will become the subspace treasury account

This way we are not requiring admins to pay with their own balance but, instead, we are allowing all admins that have a permission, to create grant on behalf of the treasury account. Which should be the one effectively paying for the various fees

Reply:
It is interesting idea, but it will allow unexpected behaviour, which is, currently the subspace manager(s) can assign any accounts as `treasury`, then the victim account will pay the fees for the subspace. 

In order to solve the issue mentioned above, we need a another proof that the subspace has the access to control the treasury account, it would another implementation something like chain link but for subspace. The process will be like:
1. treasury account sign a proof whose target is subspace id
2. create a post with the proof
3. make a subspace link for the account
But it might have a another issue making contract as `granter` impossible since no one has the contract private key to build a link then grant users fee allowances.

Therefore, I prefer the current implementation to make sure the granter can control the account to pay for users.

Reply:
I see, that's right. 

What if, instead of allowing any account to be the threasury, we generate the treasury account using the `authtypes.NewModuleAddress` method? This should allow to generate a unique address based on the subspace id theoretically. We then just need to delete the ability to later edit the `treasury` field of a subspace, and we would be sure that account is 100% controlled by the subspace. 

Then, to allow spending money from the treasury, we implement a method that allows the subspace *owner* to issue grants on behalf of the treasury to perform various operations (i.e. `SpendAuthorization`).

Reply:
Yeah, it would be the best solution. In conclusion, we need a new treasury account which has the following features:
1. A unique address generated by subspace id
2. To allow granted manager(s) have permission to spend it

I think we should pause this subspace grant ADR implementation then creating a new ADR for the new treasury account.

Reply:
I made a PR to work with the new feature. https://github.com/desmos-labs/desmos/pull/1062

Reply:
Completely agree with starting another ADR to work on a better treasury address, and then coming back to this. I would mark this as _Draft_ in the meanwhile